edb/pgsql/metaschema.py

```diff
@@ -2099,22 +2168,26 @@ class StringSliceImplFunction(dbops.Function):
         SELECT
             edgedb._substr(
                 val,
-                edgedb._normalize_array_index(
-                    start, edgedb._length(val)),
-                edgedb._normalize_array_index(
-                    stop, edgedb._length(val)) -
-                edgedb._normalize_array_index(
-                    start, edgedb._length(val))
+                pg_start,
+                pg_end - pg_start
             )
+        FROM (SELECT
+            GREATEST(0, edgedb._int_bound(edgedb._normalize_array_index(
+                start, edgedb._length(val)
+            ))) as pg_start,
+            GREATEST(0, edgedb._int_bound(edgedb._normalize_array_index(
```

Comment:
Why is `GREATEST(` needed here?

Reply:
Even though both bounds are int, when you subtract them to get length, it may overflow. 

This can happen, for example if you use `[1, 2, 3][-2147483648:12]`. Start will get converted into `-2147483648 + 3 + 1 = -2147483644`. Length would be `12 - (-2147483644) = overflow`.

This was my afternoon today.